### PR TITLE
Add Async classes to typing stub.

### PR DESCRIPTION
**Pull request by rowillia**

### CODE REVIEW

**Comment by gvanrossum** [comment]
This is nice, but we first need to have the corresponding classes added to the non-stub typing.py. That code is in the CPython repo but also has its own little "upstream" repo: https://github.com/ambv/typehinting -- can you submit a PR for that repo too? (The typing.py module and its test are in the src directory; there python2 backport should not get this. But the added classes should be added to the PEP text in that repo too.)

File: stdlib/3/typing.pyi [hunk 1]
```diff
@@ -87,6 +87,19 @@ class Hashable(metaclass=ABCMeta):
...
+    def __await__(self) -> Iterator[_T_co]:...
+
+class AsyncIterable(Generic[_T_co]):
+    @abstractmethod
+    def __anext__(self) -> Awaitable[_T_co]:...
+
+class AsyncIterator(AsyncIterable[_T_co], Generic[_T_co]):
+    @abstractmethod
+    def __anext__(self) -> Awaitable[_T_co]:...
+    def __aiter__(self) -> 'AsyncIterator'[_T_co]:...
```
Comment (JukkaL) [inline stdlib/3/typing.pyi]: The quote should include `[]`, i.e. `'AsyncIterator[_T_co]'`.

File: stdlib/3/typing.pyi [hunk 2]
```diff
@@ -94,7 +107,7 @@ class Iterable(Generic[_T_co]):
 class Iterator(Iterable[_T_co], Generic[_T_co]):
     @abstractmethod
     def __next__(self) -> _T_co: ...
-    def __iter__(self) -> Iterator[_T_co]: ...
+    def __iter__(self) -> 'Iterator'[_T_co]: ...
```
Comment (JukkaL) [inline stdlib/3/typing.pyi]: This is a similar quoting issue.

**Comment by rowillia** [comment]
@vlasovskikh Thanks for the feedback!  I was thinking the API would be:

```py
async def f(x: int) -> Awaitable[int]:
    await asyncio.sleep(x)
    return x
```

Similar to Hack's syntax - http://docs.hhvm.com/manual/en/hack.async.php

Would that be workable with this implementation?

File: stdlib/3/typing.pyi (hunk 1 above)
Comment (rowillia) [inline stdlib/3/typing.pyi]: Done

File: stdlib/3/typing.pyi (hunk 2 above)
Comment (rowillia) [inline stdlib/3/typing.pyi]: Done

**Comment by rowillia** [comment]
@gvanrossum Coorisponding typehinting PR is https://github.com/ambv/typehinting/pull/168

**Comment by vlasovskikh** [comment]
@rowillia This makes sense. Although https://github.com/ambv/typehinting/issues/119 is still a valid request that proposes a shorter syntax.

**Comment by rowillia** [comment]
@vlasovskikh totally.  AFAICT that's the syntax PyCharm is using right now, right?

**Comment by vlasovskikh** [comment]
@rowillia Yes, we've adopted this syntax as an experiment since there are no awaitable/coroutine types right now in the typing module.

**Comment by gvanrossum** [comment]
Thanks! Merged this and the typehinting changes.

**Comment by gvanrossum** [comment]
(And fixed by a4c268c4e4b2e197670dc204cf02404db47fcfcf.)

**Pull request merged**